File: src/test/js/step-definitions/google_search_step_defs.js
const {Builder, By, Key} = require('selenium-webdriver');
const {Given, When, Then, AfterAll, After} = require('@cucumber/cucumber');
const assert = require('assert');
const until = require('selenium-webdriver/lib/until');
const And = Given;

const VIDEO_RESULTS = By.css("[jsname='TFTr6");
const VIDEO_TITLE = By.css(".WZIVy>.uOId3b>.fc9yUc>.cHaqb");
const VIDEO_LINK = By.className("X5OiLe");
const SEARCH_INPUT = By.name("q");
const ANSWER_SNIPPET = By.css("[data-tts='answers']");
const YOUTUBE_HOME_PAGE = "https://www.youtube.com";
const GOOGLE_HOME_PAGE = "https://www.google.com";


let driver;

And('I am on the Google Search page', {timeout: 60 * 1000}, async function () {
    driver = await new Builder().forBrowser('chrome').build();
    await driver.get(GOOGLE_HOME_PAGE);
});

When('I search with the keyword {string}', {timeout: 2 * 5000}, async function (searchTerm) {
    const element = driver.wait(until.elementLocated(SEARCH_INPUT));
    element.sendKeys(searchTerm, Key.RETURN);
});

Then('I see the answer {string} is highlighted', async function (answer) {
    const answerEl = driver.wait(until.elementLocated(ANSWER_SNIPPET));
    const actualAnswer = await answerEl.getAttribute("data-tts-text");
    assert.equal(answer, actualAnswer);
});

Then('The search result number {int} will be the song {string}', async function (result, songName) {
    // const videos = await driver.findElements(VIDEO_RESULTS);
    // this.videoSearchResult = videos[result - 1];
    // this.expectedSongName = songName;
    // const titleEl = await this.videoSearchResult.findElement(VIDEO_TITLE);
    // const title = await titleEl.getText();
    // console.log(`The song name ${title}`);
    // assert.equal(true, title.contains(songName));
});

Then('I can play it on Youtube', async function () {
    // const videoLink = await this.videoSearchResult.findElements(VIDEO_LINK).get(1);
    // videoLink.click();
    // const currentUrl = await driver.getCurrentUrl();
    // console.log(`Opened url: ${currentUrl}`);
    // assert.equal(true, currentUrl.contains(YOUTUBE_HOME_PAGE));
    //
    // const titleOfVideoPlaying = await driver.getTitle();
    // console.log(`The playing song name ${titleOfVideoPlaying}`);
    // assert.equal(true, titleOfVideoPlaying.contains(this.expectedSongName));
});

Then('I want to close browser', async function () {
    await driver.quit();
});

File: src/test/java/cf/vlvn/nthung/steps/GoogleSearchStepDefs.java
package cf.vlvn.nthung.steps;

import cf.vlvn.nthung.pagemodel.GooglePage;
import cf.vlvn.nthung.pagemodel.YoutubePage;
import cf.vlvn.nthung.shared.TestContext;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class GoogleSearchStepDefs {
    private static final Logger log = LoggerFactory.getLogger(GoogleSearchStepDefs.class);

    private final TestContext context;
    private final GooglePage googlePage;
    private final YoutubePage youtubePage;
    private WebElement videoSearchResult;
    private String expectedSongName;

    public GoogleSearchStepDefs(TestContext context,
                                GooglePage googlePage,
                                YoutubePage youtubePage) {
        this.context = context;
        this.googlePage = googlePage;
        this.youtubePage = youtubePage;

    }

    @Given("I am on the Google Search page")
    public void iAmTheGoogleSearchPage() {
        googlePage.openHomePage();
    }

    @When("I search with the keyword {string}")
    public void iAmTypeKeywordToTheSearchBox(String searchTerm) {
        WebElement searchInput = googlePage.getSearchInput();
        searchInput.sendKeys(searchTerm);
        searchInput.submit();
    }

    @Then("The search result number {int} will be the song {string}")
    public void theSearchResultNumberWillBeTheSong(int result, String songName) {
        videoSearchResult = googlePage.getVideoSearchResultAt(result);
        this.expectedSongName = songName;
        String title = googlePage.getNameFromVideoResult(videoSearchResult);
        log.debug("The song name {}", title);
        assertTrue(title.contains(songName));
    }

    @Then("I can play it on Youtube")
    public void iCanPlayItOnYoutube() {
        googlePage.openVideoSearch(videoSearchResult);
        String currentUrl = context.getDriver().getCurrentUrl();
        log.debug("Opened url: {}", currentUrl);
        assertTrue(youtubePage.isOpen(currentUrl));

        String titleOfVideoPlaying = youtubePage.getTitleOfVideoPlaying();
        log.debug("The playing song name {}", titleOfVideoPlaying);
        assertTrue(titleOfVideoPlaying.contains(expectedSongName));
    }

    @Then("I see the answer {string} is highlighted")
    public void iSeeTheAnswerIsHighlighted(String answer) {
        String actualAnswer = googlePage.getTheAnswerSnippet();
        assertEquals(answer, actualAnswer);
    }
}
